enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,461 @@
+			}]
+		}]
+	}]
+}
+```
+
+When `opm index add` is called again to add a bundle to the `amqstreams` package, the corresponding `bundle` cue struct can be appended to a fresh local copy of `package.cue`, along with a modified `package` struct derived from the already existing package json from the index. The resulting cue definition will then be exported to get the updated package json representation. 
+
+##### Validating a json representation using schema definition
+
+CUE allows for out of the box [validation](https://cuelang.org/docs/tutorials/tour/intro/validation/) of json/yaml instances of respective CUE definitions that can be leveraged to implement `opm index validate`.
+
+
+###### A quick note about CUE
+
+As mentioned above while discussing validation of indexes, a lot of the rules about the structure of an index are implicitly documented in various sources that are scattered. A best effort is being made to keep the rules updated in the form of a library in the [api repo](https://github.com/operator-framework/api/tree/master/pkg/validation) that is imported by tools like `opm`, `operator-courier` etc that perform index validation. However, like all tools that aspire to provide data validation, these tools, or api libraries that these tools import rules from, have to be constantly maintained to keep up with the latest rule changes, while also maintaining older rules in previous releases.
+
+CUE can help reduce this maintenance overhead with it's schema definitions that can encode all of these rules in a single `.cue` file, which makes them significantly more maintainable across versions.
+
+The only Caveat of CUE is that it is pre-1.0.0 and is under rapid development, and is therefore subject to breaking changes. Vendoring in CUE's libraries instead of shelling out to the `cue` tool where ever required might mitigate this issue somewhat.
+
+Another alternative to `CUE` for implementing the declarative index config could be [conftest](https://www.conftest.dev/). However, both `CUE` and `conftest` are fairly young projects. 
+
+
+### Migration Plan
+
+With the index declarative config implemented, all other components (operator-registry, dependency-resolver etc) will be able to leverage the json representation inside the indexes to query for their content, instead of using all the artifacts to build a sqllite database as it is done currently by these components. 
+
+In order to do so, existing indexes must have the json representations in them so that all indexes in the new world are in cohort and can be used by `opm` and other components the same way. The only way to achieve this would be to rebuild the current indexes. However, since the `opm index add` command will be enhanced to create the json representations whenever it's called, the migration plan would be to collect the bundles present in an index, use the existing `opm index prune`/`opm index rm` command to remove all packages, and re-add the bundles to the index. 
```

Comment:
this doesn't seem like an acceptable migration plan.  We can't break every existing catalog index that already exists, so other olm components need to keep consuming the grpc api.  They can introduce support for a new api, but they can't drop the old one.

Reply:
I agree, I think that we need a way at runtime for OLM to pull the index down and assert if it should unpack the filesystem or attempt to serve the API. Maybe it attempts to unpack the fileystem, and if it fails to do so it attemps to run it? To start with, I think we need the newer indexes to work on older clusters and newer clusters to still interact with existing index images. To do that, I think we need to serve the new index format with the binary and just have opm understand how to have a shim between the two formats. That should make this change invisible to existing users.

Eventually, (maybe in two/three releases?) we can have a world where the index image is only built as the filesystem and without the binary image, but for now we need a migration path that doesn't involve breaking any of the existing stuff in a single minor release.

Reply:
Updated with a migration plan to convert database information to package representations when the `inspect` sub-command is called. PTAL, thanks!